corehq/apps/api/odata/tests/test_service.py

```diff
@@ -140,3 +140,123 @@ def test_with_case_types(self):
+        self.assertEqual(response.status_code, 401)
+
+    def test_wrong_domain(self):
+        other_domain = Domain(name='other_domain')
+        other_domain.save()
+        self.addCleanup(other_domain.delete)
+        correct_credentials = self._get_correct_credentials()
+        response = self.client.get(
+            reverse(self.view_urlname, kwargs={'domain': other_domain.name, 'app_id': 'my_app_id'}),
+            HTTP_AUTHORIZATION='Basic ' + correct_credentials,
+        )
+        self.assertEqual(response.status_code, 403)
+
+    def test_missing_feature_flag(self):
+        correct_credentials = self._get_correct_credentials()
+        response = self._execute_query(correct_credentials)
+        self.assertEqual(response.status_code, 404)
+
+    @flag_enabled('ODATA')
+    def test_no_case_types(self):
+        correct_credentials = self._get_correct_credentials()
+        with patch('corehq.apps.api.odata.views.get_xmlns_by_app', return_value=[]):
+            response = self._execute_query(correct_credentials)
+        self.assertEqual(response.status_code, 200)
+        self.assertEqual(
+            json.loads(response.content.decode('utf-8')),
+            {"@odata.context": "http://localhost:8000/a/test_domain/api/v0.5/odata/Forms/my_app_id/$metadata", "value": []}
+        )
+
+    @flag_enabled('ODATA')
+    def test_with_case_types(self):
+        correct_credentials = self._get_correct_credentials()
+        with patch('corehq.apps.api.odata.views.get_xmlns_by_app', return_value=['xmlns_1', 'xmlns_2']):
+            response = self._execute_query(correct_credentials)
+        self.assertEqual(response.status_code, 200)
+        self.assertEqual(
+            json.loads(response.content.decode('utf-8')),
+            {
+                "@odata.context": "http://localhost:8000/a/test_domain/api/v0.5/odata/Forms/my_app_id/$metadata",
+                "value": [
+                    {'kind': 'EntitySet', 'name': 'xmlns_1', 'url': 'xmlns_1'},
+                    {'kind': 'EntitySet', 'name': 'xmlns_2', 'url': 'xmlns_2'},
+                ],
+            }
+        )
+
+
+class TestFormServiceDocumentUsingApiKey(TestFormServiceDocument):
+
+    @classmethod
+    def setUpClass(cls):
+        super(TestFormServiceDocumentUsingApiKey, cls).setUpClass()
+        cls.api_key = ApiKey.objects.get_or_create(user=cls.web_user.get_django_user())[0]
+        cls.api_key.key = cls.api_key.generate_key()
+        cls.api_key.save()
+
+    @classmethod
+    def _get_correct_credentials(cls):
+        return TestFormServiceDocumentUsingApiKey._get_basic_credentials('test_user', cls.api_key.key)
+
+
+@flag_enabled('TWO_FACTOR_SUPERUSER_ROLLOUT')
+class TestFormServiceDocumentWithTwoFactorUsingApiKey(TestFormServiceDocumentUsingApiKey):
+
+    # Duplicated because flag on inherited method doesn't work when outer flag is used
+    @flag_enabled('ODATA')
+    def test_no_case_types(self):
```

Comment:
can these tests be rewritten such that they actually share logic with the superclass as opposed to overwriting the entire test?

Reply:
yes - updated